core/chains/evm/logpoller/orm.go

```diff
@@ -196,24 +197,43 @@ func (o *DbORM) DeleteBlocksBefore(end int64, qopts ...pg.QOpt) error {
 }
 
 func (o *DbORM) DeleteLogsAndBlocksAfter(start int64, qopts ...pg.QOpt) error {
-	// These deletes are bounded by reorg depth, so they are
-	// fast and should not slow down the log readers.
 	return o.q.WithOpts(qopts...).Transaction(func(tx pg.Queryer) error {
+		// As long as LogPoller is the only writer to these tables, we don't need to worry about having stale latest block.
+		// There is a single go routine for each chain, so we can get the latest block and safely apply it
+		// as the upper bound for the delete queries using default transaction isolation level.
+		latestBlock, err := o.SelectLatestBlock(pg.WithQueryer(tx))
+		if err != nil {
+			o.lggr.Warnw("Cant get latest block for DeleteLogsAndBlocksAfter", "err", err)
+			return err
+		}
+
 		args, err := newQueryArgs(o.chainID).
 			withStartBlock(start).
+			withEndBlock(latestBlock.BlockNumber).
 			toArgs()
 		if err != nil {
 			o.lggr.Error("Cant build args for DeleteLogsAndBlocksAfter queries", "err", err)
 			return err
 		}
 
-		_, err = tx.NamedExec(`DELETE FROM evm.log_poller_blocks WHERE block_number >= :start_block AND evm_chain_id = :evm_chain_id`, args)
+		// Applying upper bound filter is critical for Postgres performance
+		// because it allows the planner to properly estimate the number of rows to be scanned.
+		// If not applied, these queries can become very slow, because after some critical number
+		// of logs, Postgres will try to scan all the logs in the index by block_number.
+		// Latency without upper bound filter can be orders of magnitude higher for large number of logs.
+		_, err = tx.NamedExec(`DELETE FROM evm.log_poller_blocks 
+       						WHERE evm_chain_id = :evm_chain_id
+       						AND block_number >= :start_block
+       						AND block_number <= :end_block`, args)
 		if err != nil {
 			o.lggr.Warnw("Unable to clear reorged blocks, retrying", "err", err)
 			return err
 		}
 
-		_, err = tx.NamedExec(`DELETE FROM evm.logs WHERE block_number >= :start_block AND evm_chain_id = :evm_chain_id`, args)
+		_, err = tx.NamedExec(`DELETE FROM evm.logs 
```

Comment:
Alternative approach would be
```go
DELETE FROM evm.logs 
       						WHERE evm_chain_id = :evm_chain_id 
       						AND block_number >= :start_block
       						AND block_number <= (SELECT MAX(block_number) FROM evm.log_poller_blocks WHERE evm_chain_id = :evm_chain_id)
```

Seems to have similar performance, the only difference is that it's more defensive because it gets latest block from `logs`. This will handle all the cases when block was not persisted for some reason. Previous implementation takes block from `evm.log_poller_blocks`

```sql
=> begin;

explain analyze DELETE FROM evm.logs
                                                WHERE evm_chain_id = 11155111
                                                AND block_number >= 5269551
                                                AND block_number <= (SELECT MAX(block_number) FROM evm.logs WHERE evm_chain_id = 11155111);

rollback;
BEGIN
                                                                                QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Delete on logs  (cost=1295.57..75840.88 rows=21790 width=6) (actual time=0.425..0.426 rows=0 loops=1)
   InitPlan 2 (returns $1)
     ->  Result  (cost=1.18..1.19 rows=1 width=8) (actual time=0.031..0.032 rows=1 loops=1)
           InitPlan 1 (returns $0)
             ->  Limit  (cost=0.56..1.18 rows=1 width=8) (actual time=0.030..0.030 rows=1 loops=1)
                   ->  Index Only Scan Backward using evm_logs_idx on logs logs_1  (cost=0.56..2681558.66 rows=4358040 width=8) (actual time=0.029..0.029 rows=1 loops=1)
                         Index Cond: ((evm_chain_id = '11155111'::numeric) AND (block_number IS NOT NULL))
                         Heap Fetches: 1
   ->  Bitmap Heap Scan on logs  (cost=1294.38..75839.69 rows=21790 width=6) (actual time=0.072..0.106 rows=87 loops=1)
         Recheck Cond: ((evm_chain_id = '11155111'::numeric) AND (block_number >= 5269551) AND (block_number <= $1))
         Heap Blocks: exact=9
         ->  Bitmap Index Scan on evm_logs_idx  (cost=0.00..1288.93 rows=21790 width=0) (actual time=0.056..0.057 rows=87 loops=1)
               Index Cond: ((evm_chain_id = '11155111'::numeric) AND (block_number >= 5269551) AND (block_number <= $1))
 Planning Time: 0.242 ms
 Execution Time: 0.460 ms
(15 rows)
```

Reply:
Yeah, I think this way is safer... and maybe even more efficient since it's only 1 network request instead of 2.